empress/taxonomy_utils.py

```diff
@@ -0,0 +1,151 @@
+    VALID_TAXONOMY_COLUMN_NAMES = ("taxon", "taxonomy")
+    lowercase_col_names = [str(c).lower() for c in feature_metadata.columns]
+
+    # See if there is a "taxonomy column", and do some related validation on
+    # column names
+    invalid_level_columns_present = False
+    tax_col_index = None
+    tax_col_name = None
+    for col, i in zip(lowercase_col_names, range(len(lowercase_col_names))):
+        if col in VALID_TAXONOMY_COLUMN_NAMES:
+            if tax_col_index is None:
+                tax_col_index = i
+                # ("col" has already been set to lowercase)
+                tax_col_name = feature_metadata.columns[i]
+            else:
+                # Error condition 1 -- multiple possible "taxonomy columns" :(
+                raise TaxonomyError(
+                    (
+                        "Multiple columns in the feature metadata have one of "
+                        "the following names (case insensitive): {}. At most "
+                        "one feature metadata column can have a name from "
+                        "that list."
+                    ).format(VALID_TAXONOMY_COLUMN_NAMES)
+                )
+        if col.startswith("level"):
+            # This will be a problem *if* there's a taxonomy column, but if
+            # there isn't it's not a problem. (So we wait until after we've
+            # seen all of the feature metadata columns to raise an error about
+            # this.)
+            invalid_level_columns_present = True
+
+    if tax_col_index is not None:
+        if invalid_level_columns_present:
+            # Error condition 2 -- there is at least one "Level" column already
+            # in the feature metadata, which will make distinguishing these
+            # column(s) from the columns we were going to add from feature
+            # metadata difficult.
+            raise TaxonomyError(
+                "The feature metadata contains a taxonomy column, but also "
+                "already contains column(s) starting with the text 'Level' "
+                "(case insensitive)."
+            )
+
+        max_sc_count = 0
+
+        def find_max_semicolon_count(fm_row):
+            # We need to be able to update max_sc_count as we go through each
+            # row, and Python gets angry if we don't explicitly declare it as
+            # "nonlocal" before referencing it here. See
+            # https://stackoverflow.com/a/46018922/10730311
+            nonlocal max_sc_count
+            row_sc_count = fm_row.loc[tax_col_name].count(";")
+            max_sc_count = max(max_sc_count, row_sc_count)
+
+        # Figure out maximum number of semicolons in any feature (this works
+        # on a per-row basis even though axis="columns", because each call
+        # to validate_semicolon_count() gets a Series of columns which is
+        # technically a row I guess)
+        feature_metadata.apply(find_max_semicolon_count, axis="columns")
```

Comment:
I think the following would be equivalent:

```suggestion
    feature_metadata[tax_col_name].str.count(';').max()
```

Reply:
Wow, now I feel silly for writing all of that out :P Thanks for the heads up!